controllers/secret/secret_controller.go

```diff
@@ -0,0 +1,66 @@
+// Copyright 2022 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package secret
+
+import (
+	"context"
+	"github.com/go-logr/logr"
+	"github.com/vmware-samples/load-balancer-operator-for-kubernetes/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"os"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// SetupWithManager adds this reconciler to a new controller then to the
+// provided manager.
+func (r *SecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		// Watch ConfigMap resources.
+		For(&corev1.Secret{}).
+		Complete(r)
+}
+
+// SecretReconciler reconciles avi-credentials and avi-ca secrets
+type SecretReconciler struct {
+	client.Client
+	Log              logr.Logger
+	Scheme           *runtime.Scheme
+	CredentialInited bool
+	CAInited         bool
+}
+
+// AKOO will reconcile default secrets avi-controller-credentials and avi-controller-ca
+// If customer uses their own secrets, AKOO has to be restarted when secrets are updated
+func (r *SecretReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("Secret", req.NamespacedName)
+
+	secret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, req.NamespacedName, secret); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Secret not found, will not reconcile")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	if secret.Name != v1alpha1.AviCredentialName && secret.Name != v1alpha1.AviCAName {
+		return ctrl.Result{}, nil
+	}
+
+	// Avoid restart during bootup
+	if r.CredentialInited && r.CAInited {
+		log.Info("avi secret " + req.Name + " updated. Shutdown ako-operator")
+		os.Exit(0)
+	}
+
+	if secret.Name == v1alpha1.AviCredentialName {
```

Comment:
Some honest thoughts:

This can work for event-driven function, but the `Reconcile` function is level-based. We shouldn't rely on reconcile function as some sort of `onUpdated` callback. Please checkout https://stackoverflow.com/questions/67472710/getting-the-type-of-event-that-triggered-a-controller-in-kubebuilder


I feel like the reconcile logic should be looks like, by using a field `original-data` as an annotation:

```
1. if there is no `original-data` annotation on the secret, add the `original-data` with current secret value
2. if `original-data` differs from the observed secret data, update `original-data` with current values and reboot the ako-operator
```

Reply:
Good points. Thanks Fang

`avi-controller-<>` secrets have the `original-data` annotation already (may be managed by kapp?). Assume user has updated the secret once. If we follow point (2) then akoo will enter a loop of restarts as akoo will always see that observed data is different than original-data. 

CredentialInited/CAInited are not tracking if the secrets are updated. They are tracking if akoo has read them once. Without this akoo will keep on restarting. I can rename the variables if its confusing.

Let me know your thoughts